### ISSUE TITLE
ORA-08177: can't serialize access for this transaction - After Oracle Upgrade to 19c

### ISSUE THREAD

**Issue by bsislow** [laurenz/oracle_fdw#401]
We're receiving the following error on _some_ foreign tables when using "analyze" in PostgreSQL against a database that was recently upgraded to 19c (19.6):

```
ERROR:  error fetching result: OCIStmtFetch2 failed to fetch next result row
DETAIL:  ORA-08177: can't serialize access for this transaction
```

**Comment by laurenz** [comment]
That's Oracle's lame implementation of `SERIALIZABLE`.

Does that happen always or only sometimes?

I find that any concurrent data modifications are likely to trigger that error in Oracle.

**Comment by bsislow** [comment]
Sometimes.

Like I mentioned, the target Oracle database was upgraded to 19.6 from 11.2.0.4. There could be an internal behavior change. This is when the issues started.

**Comment by laurenz** [comment]
You have two options:

- Use the latest development version of oracle_fdw.

  Then change the server:

      ALTER SERVER myserver OPTIONS (ADD isolation_level 'read_committed');

  Then your `ANALYZE` should go through.

  To be on the safe side, you should then reset the setting to the default `serializable`.
  If you *know* you will have only simple queries, where there is no danger that the foreign table would have to be scanned several times during a single query, you can also stick with `READ COMMITTED`.

- Reproduce the behavior with only sqlplus:

      SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;
      SELECT * FROM oracle_table;

  If you get a failure, complain to Oracle that they fix their implementation of `SERIALIZABLE` so that it actually becomes useful.

  But don't hold your breath for them to get it fixed. Complaining would still be a good idea, to make them feel your pain.

**Comment by bsislow** [comment]
Is this version stable enough to use in production and/or when will a stable release be provided?

**Comment by laurenz** [comment]
Both.

I will probably build a new release when v13 comes out, but I consider the development version stable.

**Comment by bsislow** [comment]
We will install it and verify success shortly...

**Comment by bsislow** [comment]
Our sysadmins prefer to install a full RPM instead of from source - so we will only be able to verify this when the next stable release is available.

**Comment by laurenz** [comment]
There is no problem with rolling an RPM from the development version, but sure.

**Comment by bsislow** [comment]
Hello - has the next version been officially released? Thanks.

**Comment by laurenz** [comment]
No, this will probably happen this fall.

**Comment by laurenz** [comment]
I released version 2.3.0 today.
